ark/phenotyping/som_utils.py

```diff
@@ -67,6 +68,136 @@ def compute_cluster_avg(fovs, channels, base_dir, cluster_col,
+            should be 'cluster' or 'hCluster_cap'
+
+    Returns:
+        pd.DataFrame:
+            cell x cluster list of counts of each pixel SOM/meta cluster per each cell
+    """
+
+    # define the counts per segmentation label DataFrame
+    cluster_counts = pd.DataFrame()
+
+    # read the cell table data
+    cell_table = pd.read_csv(cell_table_path)
+
+    # subset on fov, label, and cell size
+    cell_table = cell_table[['fov', 'label', 'cell_size']]
+
+    # for verification purposes, make sure the fovs are sorted in numerical order
+    fovs_sorted = sorted(fovs, key=lambda x: int(re.findall(r'\d+', x)[0]))
+
+    for fov in fovs_sorted:
+        # read the full dataset with pixel and meta clusters
+        fov_pixel_data = feather.read_dataframe(
+            os.path.join(base_dir, consensus_dir, fov + '.feather')
+        )
+
+        # discard fov as we'll get it from the cell table
+        fov_pixel_data = fov_pixel_data.drop(columns='fov')
+
+        # group the data by segmentation label and cluster value
+        group_by_cluster_col = fov_pixel_data.groupby(
+            ['segmentation_label', cluster_col]
+        ).size().reset_index(name='count')
+
+        # cast the cluster column to int for aesthetics
+        group_by_cluster_col[cluster_col] = group_by_cluster_col[cluster_col].astype(int)
+
+        # create a pivot table with counts of each cluster per segmentation label
+        num_cluster_per_seg_label = group_by_cluster_col.pivot(
+            index='segmentation_label', columns=cluster_col, values='count'
+        ).fillna(0).astype(int)
+
+        # change column names to indicate which cluster type (pixel or meta) is used
+        new_columns = ['%s_' % cluster_col + str(c) for c in num_cluster_per_seg_label.columns]
+        num_cluster_per_seg_label.columns = new_columns
+
+        # copy over the segmentation label data
+        num_cluster_per_seg_label['segmentation_label'] = num_cluster_per_seg_label.index.values
+
+        # reset the index
+        num_cluster_per_seg_label = num_cluster_per_seg_label.reset_index(drop=True)
+
+        # subset cell table on specific fov, convert labels to int to match data types
+        cell_table_fov = cell_table[cell_table['fov'] == fov]
+        cell_table_fov['label'] = cell_table_fov['label'].astype(int)
+
+        # rename label in cell_table_fov as segmentation_label for joining purposes
+        cell_table_fov = cell_table_fov.rename(columns={'label': 'segmentation_label'})
+
+        # join the two DataFrames
+        num_cluster_per_seg_label = num_cluster_per_seg_label.merge(
+            cell_table_fov,
+            how='inner',
+            on=['segmentation_label']
+        )
+
+        # concatenate to cluster_counts
+        cluster_counts = pd.concat([cluster_counts, num_cluster_per_seg_label])
+
+    # fill empty elements with 0
+    cluster_counts = cluster_counts.fillna(0)
+
+    # remove any clusters that sum to 0
```

Comment:
@cliu72 did this result in errors downstream? When do you get a cluster with no assigned cells?

Reply:
I think the only reason I had this in my code is because I was running some sort of test where I subsetted the cells at one point, which resulted in some clusters with no cells. But in the normal case, I don't think that should ever happen.

Reply:
Ok, in that case I can remove it.